nupic/research/frameworks/dendrites/modules/dendritic_mlp.py

```diff
@@ -0,0 +1,225 @@
+
+        # Forward weight initialization must of one of "kaiming" or "modified",
+        # "hardcoded"
+        assert dendrite_init in ("kaiming", "modified", "hardcoded")
+
+        super().__init__()
+
+        self.num_segments = num_segments
+        self.input_size = input_size
+        self.hidden_size = hidden_size
+        self.output_size = output_size
+        self.dim_context = dim_context
+        self.kw = kw
+        self.hardcode_dendrites = (dendrite_init == "hardcoded")
+
+        # Forward layers & k-winners
+        self.dend1 = dendritic_layer_class(
+            module=nn.Linear(input_size, hidden_size, bias=True),
+            num_segments=num_segments,
+            dim_context=dim_context,
+            module_sparsity=0.95,
+            dendrite_sparsity=0.0 if self.hardcode_dendrites else 0.95,
+        )
+        self.dend2 = dendritic_layer_class(
+            module=nn.Linear(hidden_size, hidden_size, bias=True),
+            num_segments=num_segments,
+            dim_context=dim_context,
+            module_sparsity=0.95,
+            dendrite_sparsity=0.0 if self.hardcode_dendrites else 0.95,
+        )
+        self.classifier = SparseWeights(module=nn.Linear(hidden_size, output_size),
+                                        sparsity=0.95)
+
+        if kw:
+
+            print(f"Using k-Winners: 0.05 'on'")
+            self.kw1 = KWinners(n=hidden_size, percent_on=0.05, k_inference_factor=1.0,
+                                boost_strength=0.0, boost_strength_factor=0.0)
+            self.kw2 = KWinners(n=hidden_size, percent_on=0.05, k_inference_factor=1.0,
+                                boost_strength=0.0, boost_strength_factor=0.0)
+
+        if weight_init == "modified":
+
+            # Scale weights to be sampled from the new inititialization U(-h, h) where
+            # h = sqrt(1 / (weight_density * previous_layer_percent_on))
+            self._init_sparse_weights(self.dend1, 0.0)
+            self._init_sparse_weights(self.dend2, 0.95 if kw else 0.0)
+            self._init_sparse_weights(self.classifier, 0.95 if kw else 0.0)
+
+        if dendrite_init == "modified":
+            self._init_sparse_dendrites(self.dend1, 0.95)
+            self._init_sparse_dendrites(self.dend2, 0.95)
+
+        elif dendrite_init == "hardcoded":
+
+            # Dendritic weights will not be updated during backward pass
+            for name, param in self.named_parameters():
+                if "segments" in name:
+                    param.requires_grad = False
+
+    def forward(self, x, context):
+        output = self.dend1(x, context=context)
+        output = self.kw1(output) if self.kw else output
+
+        output = self.dend2(output, context=context)
+        output = self.kw2(output) if self.kw else output
+
+        output = self.classifier(output)
+        return output
+
+    # ------ Weight initialization functions
+    def _init_sparse_weights(self, m, input_sparsity):
```

Comment:
Why start the function names of `_init_sparse_weights` and `_init_sparse_dendrites` with underscore?

Reply:
Also, I don't see any benefit to having these as methods instead of functions, since they never refer to `self`. We might as well keep them as stand-alone functions.

Reply:
The underscore is a python convention - it refers to the fact that it's an internal only method (external code shouldn't call this method). Since it isn't using self, I can make it a static method. It should be inside the class however because it is accessing details about the class that external callers shouldn't access.